### DISCUSSION TITLE
Raspberry Pi CANopen and LINAK A25 actuator

### DISCUSSION THREAD

**Discussion by N00BMIN3R** [canopen-python/canopen#351] [Q&A]
Dear readers,

For my undergraduate electrical engineering graduation project I will need to get a Raspberry Pi 3 model b+ to communicate with a LINAK LA25 (eventually 4 to control some lift for Autonome Mobile Robots, but let's stick with just one for sake of simplicity) actuator using CANopen. The past few weeks I have learned as much as possible on my own about CANopen since its very new to me. To make communication possible I have bought the PiCAN2 DUO.

Then I found this wonderfull CANopen library for Python full of function and features that I need to make communication happen. So let's get to it. The first thing I want to get working is being able to move the actuator in different positions (with an RPDO), and get a read-out of this position (with an TPDO). Once I get this working, I will a basis for further expansion of the program.

First things first, let me tell you something about the actuator **ON NODE ID 32** The actuator comes with a handy CANopen manual that I have added in the attachement (techline-canopen-user-manual-eng). Within this document there are 2 tables (0x2000 and 0x2001).

![image](https://user-images.githubusercontent.com/34843323/225348616-7a74c634-560e-452a-8df7-1f3f407fdbbb.png)
and
![image](https://user-images.githubusercontent.com/34843323/225348741-76ef5927-843b-4fc5-9f37-1b0896fb83f4.png)

I also saw the following table which describes a starting procedure which is needed before PDO's can be send:
![image](https://user-images.githubusercontent.com/34843323/225350354-ebdf8184-c49e-4fcb-a8af-c715e19322a4.png)

And also received an EDS-file from LINAK. Which I also have added to this plead for help.

I decided to start replicating the starting procedure first. I programmed the following: (See attachment boot-up.py).
Which resulted in the following CAN-log: (see CAN-log-start-procedure.txt). 
Lets walkthrough that startup procedure.

1. **Master Heartbeat**: I am not sure what this means? Does it refer to the Consumer or Producer heartbeat? I can request these at 0x1016 and 0x1017 (or so I have I seen in the EDS-file and documentation) with and SDO. But these do not match with CANID 701. So it might be something else?
2. **Actuator Boot-up**: This I am able to see in the candump when I plugin my actuator to its power supply.
3. **Receive Service Data Object** I dont know for sure if this a specific SDO. For now I have assumed this could be any and requested the vendorID with an SDO.
4. **Actuator Response:** I get a response with the same CANID, but the data is different (most likely because of the previous step)
5. **Network management NMT start:** I broadcast the NMT “OPERATIONAL” the CANID and data match.
6. **TPDO1 STOP command needed:** The actuator did not send me anything. So I requested 0x2001 subindex 4 and received the value '0'. Which means no errors according to the documentation.
7. **Stop command:**. Seeing that the ID is 220 in the example. I need to send a PDO. I don’t know for sure which stop command it needs. So I send the 0x2000 subindex 1 along with the value of 64259 (command stop actuator).

So clearly I am doing stuff wrong, and I want to fix that. This is the first where I need this communities help with.

In the mean time I have continued on a bit of code (see CANopen Raspberry & LA25) that reads-out the Position, Current and Speed with TPDO's which works without executing the starting procedure properly? Seems a bit strange to me. But to position read-out for instance is correct.
However when I send RPDO's I do get an emergency error back. I have also added the CAN-bus log of that **(NOTE: the NODE ID is 1 here)**. See CAN-bus-log.txt.

In summary:
-	What is Master Heartbeat and how do I implement it?
-	Which SDO should be send during the starting procedure?
-	Is the way I understand and use startup procedure correct?
-       Why is my actuator not willing to execute RPDO's? 
I tried to make my question as compact as I could, if things are not clear please let me know! I will clarify as much as I can for you!

Thankyou in advance for your time dear reader :)

[techline-canopen-user-manual-eng.pdf](https://github.com/christiansandberg/canopen/files/10981196/techline-canopen-user-manual-eng.pdf)
[Boot-up.zip](https://github.com/christiansandberg/canopen/files/10981339/Boot-up.zip)
[CAN-log-start-procedure.txt](https://github.com/christiansandberg/canopen/files/10981348/CAN-log-start-procedure.txt)
[LINAK-actuator-v3-1.zip](https://github.com/christiansandberg/canopen/files/10981444/LINAK-actuator-v3-1.zip)
[CANopen Raspberry & LA25v0.1.zip](https://github.com/christiansandberg/canopen/files/10981596/CANopen.Raspberry.LA25v0.1.zip)
[CAN-bus-log.txt](https://github.com/christiansandberg/canopen/files/10981740/CAN-bus-log.txt)

**Comment by acolomb** [comment]
Hello @N00BMIN3R,

sorry your post didn't get a reply yet.  It's quite a bit of information to parse and I guess most people (including me) cannot afford the time to dive deeply into your code and troubles.  Try to read up some more on CANopen, maybe by studying other device manuals that show how the protocol is intended to be used in practice.  The one you have from LINAK sounds pretty confusing to me, mixing up several concepts and misleading about the general CANopen idea.

From a quick glance, in general it seems the PDOs are pre-configured for this device, but it doesn't say what triggers the TPDO.  Usually you send a SYNC object from the master regularly and that triggers every device to report its state via TPDO.  But there are other transmission types.  If the device really behaves like a CANopen device, you should be able to read out these details live from the Object Dictionary: `node.pdo.read()`, then watch the log for what the library finds on the device.

> 1. **Master Heartbeat**: I am not sure what this means? Does it refer to the Consumer or Producer heartbeat? I can request these at 0x1016 and 0x1017 (or so I have I seen in the EDS-file and documentation) with and SDO. But these do not match with CANID 701. So it might be something else?

That is misleading and totally unrelated.  Whether your master (Node ID 1 here) sends a Heartbeat object or not should not matter to the device at all.  It doesn't even have a concept of a "master" node, but rather all devices by default communicate with objects involving their own Node ID (720, 620, 5A0, etc.).  Just disregard this entry.

> 2. **Actuator Boot-up**: This I am able to see in the candump when I plugin my actuator to its power supply.

Nice, so it's behaving properly, and says it's in Initialization state.  No PDOs "exist" (in CANopen parlor) unless you switch to the Operational state, so it will react to its RPDO and transmit its TPDO.

> 3. **Receive Service Data Object** I dont know for sure if this a specific SDO. For now I have assumed this could be any and requested the vendorID with an SDO.

You shouldn't need to do that, it's just an example reading the Consumer Heartbeat Time (0x1016:01) object.  What would be a sensible exchange at that point is reading out the PDO configuration, so the python-canopen library knows about the data mapping and can provide you with meaningful names to access the PDO contents.  That'll be a whole bunch of SDO exchanges though, during the `pdo.read()` call.

> 5. **Network management NMT start:** I broadcast the NMT “OPERATIONAL” the CANID and data match.

*If* you have the Producer Heartbeat Time object configured to some interval, you should get 0x720 objects regularly where you can observe the state change to Operational (05).

> 6. **TPDO1 STOP command needed:** The actuator did not send me anything. So I requested 0x2001 subindex 4 and received the value '0'. Which means no errors according to the documentation.

This is where the documentation should be clear under which conditions a TPDO is sent.  Cyclic, caused by the SYNC object reception?  Asynchronous in regular intervals?  Whenever one of the involved values changes?

Try to investigate the active PDO configuration, especially the Transmission Type entry.  These are standardized modes, let's hope the device adheres to the standard enough to tell you what it expects.

> 7. **Stop command:**. Seeing that the ID is 220 in the example. I need to send a PDO. I don’t know for sure which stop command it needs. So I send the 0x2000 subindex 1 along with the value of 64259 (command stop actuator).

This should be you sending the RPDO, with the contents according to the other table.  The first field there specifies a command.  Note that with PDOs, you don't specify object index and subindex anymore, it's just interpreting the data bytes according to the mapped layout.  Or more in CANopen's official terms: The mapped objects are applied to the Object Dictionary when receiving the RPDO, either immediately or upon the next SYNC object.  This again depends on the RPDO's Transmission Type configuration.

> * Which SDO should be send during the starting procedure?

Whatever you need to exchange, there is no requirement to use SDO at all if the PDOs are preconfigured.

>     * Is the way I understand and use startup procedure correct?

Apparently not, take the included table more as an example exchange, not as a recipe to follow.

Hope that helps.  If you have further questions, please try to ask as precisely as possible, with a limited scope.  It's much easier to parse and answer than a truckload of logs and example scripts.

Kind regards
André
